### ISSUE TITLE
Use go-cmp/cmp only in tests

### ISSUE THREAD

**Issue by justinsb** [kubernetes/kubernetes#104821]
I propose that we try to limit the use of go-cmp/cmp primarily to tests.

The comparisons themselves are less efficient, and referencing the cmp package may allocate some static structures.

Where we can't easily avoid calling cmp, I propose we limit its usage.  If we are calling it from a V(4) log statement (for example), we should first check whether the log level is active.

**Comment by dims** [comment]
still stuff to do here!

```
[dims@dims-m1-7728 21:07] ~/go/src/k8s.io/kubernetes ⟩ rg 'go-cmp/cmp"' | grep -v _test | grep -v vendor
pkg/registry/flowcontrol/ensurer/strategy.go:	"github.com/google/go-cmp/cmp"
pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go:	"github.com/google/go-cmp/cmp"
pkg/scheduler/profile/profile.go:	"github.com/google/go-cmp/cmp"
pkg/scheduler/framework/interface.go:	"github.com/google/go-cmp/cmp"
pkg/scheduler/apis/config/validation/validation.go:	"github.com/google/go-cmp/cmp"
pkg/apis/core/validation/validation.go:	"github.com/google/go-cmp/cmp"
test/e2e/common/node/lease.go:	"github.com/google/go-cmp/cmp"
staging/src/k8s.io/component-base/config/testing/roundtrip.go:	"github.com/google/go-cmp/cmp"
staging/src/k8s.io/component-base/config/testing/helpers.go:	"github.com/google/go-cmp/cmp"
staging/src/k8s.io/apimachinery/pkg/util/diff/diff.go:	"github.com/google/go-cmp/cmp"
staging/src/k8s.io/apimachinery/pkg/api/apitesting/roundtrip/compatibility.go:	"github.com/google/go-cmp/cmp"
staging/src/k8s.io/apiserver/pkg/storage/testing/utils.go:	"github.com/google/go-cmp/cmp"
staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:	"github.com/google/go-cmp/cmp"
staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:	"github.com/google/go-cmp/cmp"
```

**Comment by johnbelamaric** [comment]
Seems we need a linter check or something

**Comment by s-urbaniak** [comment]
raising priority on this one as the godoc for this package explicitly states (https://pkg.go.dev/github.com/google/go-cmp/cmp):
> It is intended to only be used in tests, as performance is not a goal and it may panic if it cannot compare the values. Its propensity towards panicking means that its unsuitable for production environments where a spurious panic may be fatal.

I'll open PRs that are going to try to remove the usage of this package and just rely on reflect.DeepEquals instead.

/cc @sttts

**Comment by sttts** [comment]
That reference to panics sounds more concerning than just some allocation that @justinsb mentioned in the original issue. Afaik that statement was added only recently to go-cmp.

**Comment by liggitt** [comment]
> That reference to panics sounds more concerning than just some allocation that @justinsb mentioned in the original issue. 

Agreed

> Afaik that statement was added only recently to go-cmp.

No, it's always been there: 
https://github.com/google/go-cmp/commit/1c57fff5373af756ecb7a9aa4860f55ad0782b36#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R22

**Comment by ffromani** [comment]
/triage accepted

it's maybe silly, but let's fix the labels

**Comment by dims** [comment]
@s-urbaniak are you still interested in helping with this?

**Comment by s-urbaniak** [comment]
@dims yes definitely! :-) last discussion point is https://github.com/kubernetes/kubernetes/pull/129161#discussion_r1907565324, let's wait for @thockin's final decision points, taking his guidance I am more than willing to help :)

**Comment by thockin** [comment]
Sorry, what are we waiting on me for?  We should weed it out of prod code.

**Comment by dims** [comment]
@thockin here's the linter PR! https://github.com/kubernetes/kubernetes/pull/129815

**Comment by pohly** [comment]
Is using cmp.Diff for log entries considered okay as long as it is used with a "only call if really logging at high level" pattern (as mentioned in the description) or not (as @dims said in https://github.com/kubernetes/kubernetes/pull/129032#issuecomment-2614158742)?

**Comment by pohly** [comment]
It would be possible to provide a helper which enables something like this:
```
logger.V(4).Info("hello world", "diff", klog.AsStringer(func() string { return cmp.Diff(...) }))
```

`klog.AsStringer` wraps the function as a `fmt.Stringer` and only calls it if the value is really being logged. It can catch panics, just in case that a logging backend doesn't do that by itself (our current ones do).

**Comment by pgimalac** [comment]
For the record https://github.com/google/go-cmp 's `Equal` and `Diff` functions also disable some dead code elimination due to using `reflect.Type.Method` (if you use it, then every public method of every reachable type has to be kept by the linker, making your binaries bigger), see https://github.com/google/go-cmp/issues/373 for details.
In my case k8s using `go-cmp` is one of the very few reasons the optimization is disabled in my binaries !

**Comment by pohly** [comment]
I take that as a "no, don't even use go-cmp for logging". Some folks will be unhappy about that, I've seen quite a few places where people really wanted that to make the log output useful to them for at debug log levels.

**Comment by pohly** [comment]
We could also do a `<loghelperpackage>.Diff(a, b any) fmt.Stringer` implementation which calls go-cmp under the hood. Calling that would be allowed by linting, calling go-cmp directly remain forbidden.

But that then means that we keep having that [dead code elimination problem](https://github.com/kubernetes/kubernetes/issues/104821#issuecomment-2668697492). We also don't have a good place to put such a `Diff`. `klog` can't take the `go-cmp` dependency, `component-base` is too large.

**Comment by pohly** [comment]
We could implement an alternative with some kind of pretty-printer and a custom diff implementation for text lines. But which pretty-printer? go-spew is already a [Kubernetes dependency](https://github.com/kubernetes/kubernetes/blob/0b1ef24a9912d1cbbbfe5641bf9cd2cdc91f32bd/staging/src/k8s.io/apimachinery/pkg/util/dump/dump.go#L20), but was last updated in 2018. I'd rather remove it, but that's not easy (required for checksuming of on-disk kubelet files - don't ask!) and a different topic.

Perhaps a simpler diff for types which implement fmt.Stringer would be sufficient. Except that the generated `String` method for our API types is neither readable nor easily diffable (single line...)...

**Comment by thockin** [comment]
> I take that as a "no, don't even use go-cmp for logging".

That is what I was saying, yes.  :)

> We could implement an alternative with some kind of pretty-printer and a custom diff implementation for text lines.

I think this is the right path.  First, I will suggest that KYAML is a decent pretty-print format (if we add unexported-fields and cycle detection).  We compartmentalized spew so that we could excise it or reimplement it or fork it into k-sigs.  Hypothetically, if we implemented kspew which dumped to a format we like with code we understand (spew is pretty complex), and then had a simple unified diff (simpler than cmp.Diff) function, doesn't that solve a lot?

**Comment by pohly** [comment]
KYAML = https://pkg.go.dev/sigs.k8s.io/kustomize/kyaml ?

> Hypothetically, if we implemented kspew which dumped to a format we like with code we understand (spew is pretty complex), and then had a simple unified diff (simpler than cmp.Diff) function, doesn't that solve a lot?

Yes. We just need someone to do this and then maintain it. We forked go-yaml and then [no-one looked after issues](https://github.com/kubernetes-sigs/yaml/issues/117#issuecomment-2564082275).

**Comment by pohly** [comment]
Another alternative for a pretty-printer is https://pkg.go.dev/github.com/onsi/ginkgo/v2/formatter. We would need to fork it, of course, to avoid the full Ginkgo dependency.

**Comment by pohly** [comment]
Scratch that, I meant https://pkg.go.dev/github.com/onsi/gomega/format.

**Comment by thockin** [comment]
I meant KYAML = https://github.com/kubernetes/kubernetes/pull/130199 which is (not surprisingly) similar to every JSON-ish encoder out there.

**Comment by thockin** [comment]
> We forked go-yaml and then no-one looked after issues.

Yes, having looked at that code a lot lately, I think we made a strategic error.  I'm not sure what the better path would be, but that code is INCREDILY twisty and convoluted, converting to and from JSON, with v2 and v3 libraries, but mostly using v2, resulting in bad error messages and a non-standards-compliant decoder (at least I could not make some things work).

YAML is a disaster overall and I don't know how to get out of the mess.  That's why my kyaml PR has a from-scratch encoder (note: encoding is WAY easier than decoding :) rather than try to teach our YAML lib to be more flexible about encoding.  It's just a lost cause.

**Comment by liggitt** [comment]
> We forked go-yaml and then [no-one looked after issues](https://github.com/kubernetes-sigs/yaml/issues/117#issuecomment-2564082275).

The reason we forked it was to maintain compatibility with k8s use in yaml.v3 and be able to address security issues. I don't think there's any intent to try to address long-lived not-very-impactful bugs in the yaml library.

**Comment by pohly** [comment]
> I saw that issue get reported, but didn't prioritize it because it didn't look very impactful to kubernetes.

True. But to me it looked it wasn't getting any attention.

Forking is okay if done carefully and planning for contingencies. We should not rely on someone saying "I'll maintain it". Such repos need to be looked after by a SIG, including periodic triaging of issues. If some maintainer wants to or has to step down, someone else has to be found.

We cannot rely on Jordan (or Tim, or Patrick, or some of the other usual suspects) to go in and fix things (like Jordan did for go-yaml).

**Comment by thockin** [comment]
Not to make light of it, but if we need the code, we need the code.  Whether it lives in k/k in k-sigs or somewhere else, we need it.  We often don't need ALL of it, so we make tactical decisions about what burden is worth carrying ourselves.

Having looked at the yaml code, I tried to untangle it and gave up.  The underlying code is twisty, and I don't pretend to know all of the sublte issues around encode/decode that come from years of piecemeal evolution.  We need a better solution, but we are not able/willing/ready to rewrite it (or to own in deeply enough to make it manageable).

There is a (small and mostly theoretical) advantage to doing it OUTSIDE of kubernetes-space -- it MIGHT grow maintainers who find it via other projects.  E.g. if go-yaml is irredeemable, making a more manageable library might be better -- we are not alone in our suffering, I am sure.

**Comment by pohly** [comment]
>  Whether it lives in k/k in k-sigs or somewhere else, we need it.

Agreed. All I am saying is that when we do something like this, we should make sure that it's set up such that the responsible SIG is aware of the sub-project and takes care of it. The advantage of forking into Kubernetes, besides the ability to get changes merged of course, is that we don't need to depend on individuals.

At least in theory. In practice we of course still need people to do the work. 😢

**Comment by BenTheElder** [comment]
> I meant KYAML = https://github.com/kubernetes/kubernetes/pull/130199 which is (not surprisingly) similar to every JSON-ish encoder out there.

If we do this, we should drop go-cmp from sigs.k8s.io/yaml deps :-)

https://github.com/kubernetes-sigs/yaml/blob/56d672052dcff7362af5b66f6424976539cddd78/go.mod#L6C20-L6C26

**Comment by pohly** [comment]
I suspect that some design choices in KYAML, in particular the "eschews significant whitespace", might make the output less readable than YAML could be otherwise: long strings with embedded quotation marks and newlines are very hard to read as quoted strings, which is what KYAML uses (`fmt.Fprintf(buf, "%q", val)`). They are much more readable when YAML multiline is used. But KYAML probably won't support that because it would make whitespace significant?

Having a single line in the output for a long multi-line string value also makes a line-based diff less useful.

In klog, we already have [`klog.Format`](https://pkg.go.dev/k8s.io/klog/v2#Format) which uses pretty-printed JSON as replacement for the protobuf-generated `String` implementation. This has the same problem with string quoting, but is something which is usable already now. Would a `klog.Diff(oldObj, newObj any)` make sense? It could produce multi-line strings using the same code as `klog.Format`, then produce a diff using a hand-written code (i.e. no external go-cmp dependency!).

If we later arrive at some Kubernetes-maintained pretty-printer, we could swap out the JSON format in klog for something nicer, if the dependency is considered acceptable for klog.

**Comment by thockin** [comment]
I only mentioned kyaml because I slightly dislike JSON, but yes JSON would be fine.

cmp.Diff tries to be TOO clever, IMO, and can end up making micro diffs.  Having a simple log-diff function (could be basically a typed type of any, any) and printing that as a line diff shouldn't be TOO hard?

**Comment by pohly** [comment]
> Having a simple log-diff function (could be basically a typed type of any, any) and printing that as a line diff shouldn't be TOO hard?

Yes, that's exactly what I was proposing with `klog.Diff`. So I take that as a tentative +1 - now where's the time to implement this...
